sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/BaseScriptTransformationSuite.scala

```diff
@@ -0,0 +1,227 @@
+          script = "cat",
+          output = Seq(AttributeReference("a", StringType)()),
+          child = ExceptionInjectingOperator(child),
+          ioschema = noSerdeIOSchema
+        ),
+        rowsDf.collect())
+    }
+    assert(e.getMessage().contains("intentional exception"))
+    // Before SPARK-25158, uncaughtExceptionHandler will catch IllegalArgumentException
+    assert(uncaughtExceptionHandler.exception.isEmpty)
+  }
+
+  test("SPARK-25990: TRANSFORM should handle different data types correctly") {
+    assume(TestUtils.testCommandAvailable("python"))
+    val scriptFilePath = getTestResourcePath("test_script.py")
+    case class Struct(d: Int, str: String)
+    withTempView("v") {
+      val df = Seq(
+        (1, "1", 1.0, BigDecimal(1.0), new Timestamp(1),
+          new Date(2020, 7, 1), new CalendarInterval(7, 1, 1000), Array(0, 1, 2),
+          Map("a" -> 1)),
+        (2, "2", 2.0, BigDecimal(2.0), new Timestamp(2),
+          new Date(2020, 7, 2), new CalendarInterval(7, 2, 2000), Array(3, 4, 5),
+          Map("b" -> 2)),
+        (3, "3", 3.0, BigDecimal(3.0), new Timestamp(3),
+          new Date(2020, 7, 3), new CalendarInterval(7, 3, 3000), Array(6, 7, 8),
+          Map("c" -> 3))
+      ).toDF("a", "b", "c", "d", "e", "f", "g", "h", "i")
+          .select('a, 'b, 'c, 'd, 'e, 'f, 'g, 'h, 'i, struct('a, 'b).as("j"))
+      // Note column d's data type is Decimal(38, 18)
+      df.createTempView("v")
+
+      assert(spark.table("v").schema ==
+        StructType(Seq(StructField("a", IntegerType, false),
+          StructField("b", StringType, true),
+          StructField("c", DoubleType, false),
+          StructField("d", DecimalType(38, 18), true),
+          StructField("e", TimestampType, true),
+          StructField("f", DateType, true),
+          StructField("g", CalendarIntervalType, true),
+          StructField("h", ArrayType(IntegerType, false), true),
+          StructField("i", MapType(StringType, IntegerType, false), true),
+          StructField("j", StructType(
+            Seq(StructField("a", IntegerType, false),
+              StructField("b", StringType, true))), false))))
+
+      // Can't support convert script output data to ArrayType/MapType/StructType now,
+      // return these column still as string
+      val query = sql(
+        s"""
+           |SELECT
+           |TRANSFORM(a, b, c, d, e, f, g, h, i, j)
+           |USING 'python $scriptFilePath'
+           |AS (
+           |  a int,
+           |  b string,
+           |  c double,
+           |  d decimal(1, 0),
+           |  e timestamp,
+           |  f date,
+           |  g interval,
+           |  h string,
+           |  i string,
+           |  j string)
+           |FROM v
+        """.stripMargin)
+
+      checkAnswer(query, identity, df.select(
+        'a, 'b, 'c, 'd, 'e, 'f, 'g,
+        'h.cast("string"),
+        'i.cast("string"),
+        'j.cast("string")).collect())
+    }
+  }
```

Comment:
@maropu
 I change this UT to show that , now, we can handle input/output data. Later I will update pr desc to declare this point